src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sstring.hh>
+
+#include <memory>
+#include <vassert.h>
+
+namespace ss = seastar;
+
+namespace util {
+
+/// Tracks allocations and deallocations of memory. A mem_tracker can have
+/// child trackers effectively forming an n-ary tree like structure.
+class mem_tracker {
```

Comment:
Nothing to change, but a future feature enhancement would be to keep a histogram of allocation/deallocation sizes.  It could be useful to see if the memory consumed is a lot of small allocations or a few large ones. Even having a few buckets for powers of 2 would be enough to gain some insight into this.

Reply:
Having 16 buckets with something like 'std::array<uint8_t, 16>` and approximate counting for each bucket would limit the memory footprint of the histogram to 128bits.

Reply:
Good idea.